instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt

```diff
@@ -116,8 +179,121 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
+     * That was happening in legacy versions of NN.
+     */
+    @Test
+    fun additional_alternative_is_not_force_to_invoke_alternatives_observer() {
+        // Prepare with alternative routes.
+        setupMockRequestHandlers(coordinates)
+        val routes = requestDirectionsRouteSync(coordinates)
+
+        // Play primary route
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
+            mockLocationReplayerRule.playRoute(routes.first())
+            mapboxNavigation.startTripSession()
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            // infinity subscription to avoid triggering NN on every new observer
+            mapboxNavigation.registerRouteAlternativesObserver(
+                object : NavigationRouteAlternativesObserver {
+                    override fun onRouteAlternatives(
+                        routeProgress: RouteProgress,
+                        alternatives: List<NavigationRoute>,
+                        routerOrigin: RouterOrigin
+                    ) = Unit
+
+                    override fun onRouteAlternativesError(error: RouteAlternativesError) = Unit
+                }
+            )
+            mapboxNavigation.setRoutes(routes)
+        }
+
+        // Subscribing for alternatives
+        val firstAlternative = RouteAlternativesIdlingResource(
+            mapboxNavigation
+        ) { _, alternatives, _ ->
+            alternatives.isNotEmpty()
+        }
+        firstAlternative.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("alternatives failed") {
+            Espresso.onIdle()
+        }
+        firstAlternative.unregister()
+
         assertNotNull(firstAlternative.alternatives)
+
+        val nextAlternativeObserver = RouteAlternativesIdlingResource(
+            mapboxNavigation
+        ) { _, alternatives, _ ->
+            alternatives.isNotEmpty()
+        }
+
+        val externalAlternatives = DirectionsResponse.fromJson(
+            readRawFileText(activity, R.raw.route_response_alternative_continue)
+        ).routes().also {
+            assertEquals(1, it.size)
+        }
+        lateinit var setRoutes: List<DirectionsRoute>
+        runOnMainSync {
+            setRoutes = (
+                mutableListOf(
+                    mapboxNavigation.getRoutes().first()
+                ) + firstAlternative.alternatives!! + externalAlternatives
+                ).also {
+                assertTrue(
+                    "Primary route + >=1 alternatives + external alternatives",
+                    it.size >= 3
+                )
+            }
+            mapboxNavigation.setRoutes(setRoutes)
+        }
+
+        nextAlternativeObserver.register()
```

Comment:
Ir seems like RoutesAlternativeObserver is not sticky. I think a race is possible here: alternative observers will be notified before this one is registered.

Reply:
that's actually mustn't, as I added tests for that `alternative_routes_observer_is_not_triggered_for_set_routes` (set route is not trigger observer), but I'll move `nextAlternativeObserver.register()` above 👍